config: use `enable-timestamp` and `enable-error-stack` for configuration

ichn-hu opened the pull request:
<!--
Thank you for contributing to TiDB! Please read TiDB's [CONTRIBUTING](https://github.com/pingcap/community/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.
-->

### What problem does this PR solve? <!--add issue link with summary if exists-->

This PR solves #12748 

### What is changed and how it works?

By creating a `nullableBool` type, we allow user to use `enable-xx` while keep `disable-xx` still valid. Due to the limitation of the implementation of `Config.Load` which overrides the caller, we can't distinguish user's option with the default ones, however using `nullableBool` to replace `bool` enables us to know what the user has put in the configuration file.

### Check List <!--REMOVE the items that are not applicable-->

Tests <!-- At least one of them must be included. -->

 - Unit test
 - Integration test

Code changes

 - Has exported variable/fields change

Side effects

 - Increased code complexity

Related changes

 - Need to update the documentation

Release note

Deprecated `disable-timestamp` and `disable-error-stack` for log configuration, introduce `enable-timestamp` and `enable-error-stack` for replacement.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/pingcap/tidb?pullRequest=12853) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/pingcap/tidb?pullRequest=12853) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/pingcap/tidb?pullRequest=12853) it.</sub>

SunRunAway commented:
/run-unit-test

SunRunAway commented on config/config_test.go:
```diff
@@ -34,6 +35,91 @@ func TestT(t *testing.T) {
+	var nb = nullableBool{false, false}
+	data, err := json.Marshal(nb)
+	c.Assert(err, IsNil)
+	err = json.Unmarshal(data, &nb)
```
Do not test the data marshaling from the struct `nullableBool`, it will never happen in real world.
Here's a real-world test.
```go
data := []byte("true")
json.Unmarshal(data, &nb)
```

SunRunAway commented on config/config_test.go:
```diff
@@ -34,6 +35,91 @@ func TestT(t *testing.T) {
+		conf = defaultConf
+		_, err = f.WriteString(confStr)
+		c.Assert(err, IsNil)
+		c.Assert(f.Sync(), IsNil)
```
This line is useless.

ichn-hu replied:
You mean
```
c.Assert(f.Sync(), IsNil)
```
?

I could remove it to make you happy, but it is also borrowed from previous code.

```
f, err := os.Create(configFile)
	c.Assert(err, IsNil)

	// Make sure the server refuses to start if there's an unrecognized configuration option
	_, err = f.WriteString(`
unrecognized-option-test = true
`)
	c.Assert(err, IsNil)
	c.Assert(f.Sync(), IsNil)

	c.Assert(conf.Load(configFile), ErrorMatches, "(?:.|\n)*unknown configuration option(?:.|\n)*")

```

SunRunAway replied:
You can change the previous code. :)

SunRunAway commented on config/config_test.go:
```diff
@@ -34,6 +35,91 @@ func TestT(t *testing.T) {
+	configFile = path.Join(path.Dir(localFile), configFile)
+
+	f, err := os.Create(configFile)
+	c.Assert(err, IsNil)
```
We can use `defer` here, in case some error happens before `f.Close` and `os.Remove`.

```go
defer func() {
    f.Close()
    os.Remove(f.Name())
}()
```

ichn-hu replied:
I suppose we should keep the file if some error happens so that we could look into the file to double check since this part of code is for test purpose. This piece of code is borrowed from previous code in `TestConfig`, which did not use `defer` neither.

SunRunAway replied:
It'd be better use defer here to make the testing environment clean. TiDB has hundreds of contributors who may run the test but only few of them need to keep the file to do investigation. 
When you need double check you can delete the `os.Remove` code and rerun.

ichn-hu replied:
The file is not deleted only if some test is broken, in this case manual inspection is needed (although it is not likely to be broken if the code is not changed), and you could/should manually delete that file, or rerun the test which will delete it if not thing went wrong. I agree it make sense to use defer here, I will change it if you insist.

ichn-hu commented on config/config.go:
```diff
@@ -100,17 +101,88 @@ type Config struct {
+// conflict options at the same time.
+type nullableBool struct {
+	Valid bool `toml:"valid" json:"valid"`
+	Value bool `toml:"value" json:"value"`
```
> Since we implement `UnmarshalText` and `UnmarshalJSON`, we do not need the tags for toml and json here.

Nope, tags are for `marshal`, to follow the previous practice, we want the keys to be lowercase in marshaled data, so I put the tags here.

SunRunAway replied:
In which case we need marshal `nullableBool`?

ichn-hu replied:
No, you need, in `http_request.go`, accessing `/settings` will return a JSON encoded configuration file, in this case marshal to JSON is required, although not neccessary for marshal TOML.

ichn-hu commented:
/run-unit-test

ichn-hu commented:
> LGTM

Thanks!!!

zz-jason commented on config/config.go:
```diff
@@ -100,17 +101,82 @@ type Config struct {
+// nullableBool defaults unset bool options to unset instead of false, which enables us to know if the user has set 2
+// conflict options at the same time.
+type nullableBool struct {
+	Valid bool
```
how about:
* s/Valid/IsNull/
* s/Value/IsTrue/

ichn-hu replied:
Sorry but by default we want the two fields to be false (in case the user does not put anything in the configuration file) and result in `Unset`, if you want I could make it s/Value/NotNull to be compatible with the default value.

XuHuaiyu replied:
Will `isValid` better?

XuHuaiyu commented on config/config.go:
```diff
@@ -100,17 +101,82 @@ type Config struct {
 	Format string `toml:"format" json:"format"`
-	// Disable automatic timestamps in output.
-	DisableTimestamp bool `toml:"disable-timestamp" json:"disable-timestamp"`
+	// Disable automatic timestamps in output. Deprecated.
```
Deprecated: Use EnbaleTimestamp instead.

XuHuaiyu commented on config/config.go:
```diff
@@ -100,17 +101,82 @@ type Config struct {
+	// EnableTimestamp enables automatic timestamps in log output.
+	EnableTimestamp nullableBool `toml:"enable-timestamp" json:"enable-timestamp"`
 	// DisableErrorStack stops annotating logs with the full stack error
+	// message. Deprecated.
```
ditto

zz-jason commented on config/config.go:
```diff
@@ -561,6 +650,12 @@ func (c *Config) Load(confFile string) error {
 // Valid checks if this config is valid.
 func (c *Config) Valid() error {
+	if c.Log.EnableErrorStack == c.Log.DisableErrorStack && c.Log.EnableErrorStack != nbUnset {
+		return fmt.Errorf("configuration confliction of enable-error-stack and disable-error-stack")
```
```suggestion
		return fmt.Errorf("\"enable-error-stack\" (%v) conflicts \"disable-error-stack\" (%v). \"disable-error-stack\" is deprecated, please use \"enable-error-stack\" instead.", c.Log.EnableErrorStack, c.Log.DisableErrorStack)
```

zz-jason commented on config/config.go:
```diff
@@ -561,6 +650,12 @@ func (c *Config) Load(confFile string) error {
+		return fmt.Errorf("configuration confliction of enable-error-stack and disable-error-stack")
+	}
+	if c.Log.EnableTimestamp == c.Log.DisableTimestamp && c.Log.EnableTimestamp != nbUnset {
+		return fmt.Errorf("configuration confliction of enable-timestamp and disable-timestamp")
```
ditto

XuHuaiyu commented on config/config.go:
```diff
@@ -121,6 +187,26 @@ type Log struct {
 	RecordPlanInSlowLog uint32 `toml:"record-plan-in-slow-log" json:"record-plan-in-slow-log"`
 }
 
+func (l *Log) getDisableTimestamp() bool {
```
Do we need `getEnableTimestamp` which maybe more intuitive?

ichn-hu replied:
Nope, in the internal logic we only need one option, and it is enough to have `getDisableTimestamp`. Historically we only have `DisableTimestamp`, so I created this mock function to provide a proxy for the added `EnableTimestamp` option.

SunRunAway replied:
History can be changed by your hands.

ichn-hu replied:
The internal logic is designed to support `Disable-xx`, of course I could make them all `Enable-xx` but I'm not sure about the consequences. 

![image](https://user-images.githubusercontent.com/29735669/67670633-4bee1d00-f9af-11e9-95f7-b9b9fc3f92ec.png)

If you insist on this, I could make another PR to fix this.

XuHuaiyu replied:
The internal logic is not uniform. I think use `enable-xx` is more intuitive.

XuHuaiyu commented on config/config.go:
```diff
@@ -121,6 +187,26 @@ type Log struct {
+	return !l.EnableTimestamp.toBool()
+}
+
+func (l *Log) getDisableErrorStack() bool {
```
ditto

XuHuaiyu commented on config/config.go:
```diff
@@ -361,7 +447,10 @@ var defaultConf = Config{
 		ExpensiveThreshold:  10000,
-		DisableErrorStack:   true,
+		DisableErrorStack:   nbUnset,
+		EnableErrorStack:    nbUnset, // If both options are unset, getDisableErrorStack() returns true
```
What does `both options` mean?

ichn-hu replied:
Both options refer to `DisableErrorStack` and `EnableErrorStack`, `DisableErrorStack` is kept for back compatibility, but to make it possible to detect what the user has put in the configuration file, we have to make both option unset by default, in which case `getDisableErrorStack` returns true.

XuHuaiyu commented on config/config.go:
```diff
@@ -361,7 +447,10 @@ var defaultConf = Config{
+		DisableErrorStack:   nbUnset,
+		EnableErrorStack:    nbUnset, // If both options are unset, getDisableErrorStack() returns true
+		EnableTimestamp:     nbUnset,
+		DisableTimestamp:    nbUnset, // if both options are unset, getDisableTimestamp() returns false
```
s/ if/ If

XuHuaiyu commented on config/config.go:
```diff
@@ -361,7 +447,10 @@ var defaultConf = Config{
 		ExpensiveThreshold:  10000,
-		DisableErrorStack:   true,
+		DisableErrorStack:   nbUnset,
+		EnableErrorStack:    nbUnset, // If both options are unset, getDisableErrorStack() returns true
```
s/ unset/ nbUnset

XuHuaiyu commented on config/config.go:
```diff
@@ -361,7 +447,10 @@ var defaultConf = Config{
+		DisableErrorStack:   nbUnset,
+		EnableErrorStack:    nbUnset, // If both options are unset, getDisableErrorStack() returns true
+		EnableTimestamp:     nbUnset,
+		DisableTimestamp:    nbUnset, // If both options are unset, getDisableTimestamp() returns false
```
ditto

ichn-hu commented:
> Addressing the comments make this a LGTM

Thanks for your comments, I've addressed them in the update. However for the `enable-xx` I would like to do it later in another PR. Currently the `NewLogConfig` accepts two `disable-xx`, it make sense to refactor the internal logic to support `enable-xx` fundamentally, but it might have too many consequences (since there are 26 places where `NewLogConfig` is called, spread across many files), therefore I prefer to keep the legacy logic for now but refractor later in another PR.

ichn-hu commented:
/run-unit-test
